### ISSUE TITLE
[function calling/action]: implement retry or redundancy 

### ISSUE THREAD

**Issue by louis030195** [stellar-amenities/hal-9100#93]
Depending on constraints could:
1. retry sequentially if a tool is crashing (error)
2. generate in parallel x times and pick the first that does not crash (speed optimisation)

Out of scope/overkill:
- Generate x times in parallel + classification which one is the best?

Example:

- action tool, the llm generated wrong syntax in a query param, request fails 400, retry

**Comment by louis030195** [comment]
similar to code interpreter async LLM recursion "you generated this and it failed, please make it work ...."
